voicevox_engine/guided.py

```diff
@@ -0,0 +1,338 @@
+        f0 *= predicted_avg / f0_avg
+
+    f0 *= 2 ** pitchScale
+    f0[f0 > 6.5] = 6.5
+    f0[(0 < f0) & (f0 < 3)] = 3.0
+
+    f0 = resample(f0, int(len(f0) / speedScale))
+    phone_list = resample(phone_list, int(len(phone_list) / speedScale))
+
+    return engine.guided_synthesis(
+        length=phone_list.shape[0],
+        phoneme_size=phone_list.shape[1],
+        f0=f0[:, np.newaxis].astype(np.float32),
+        phoneme=phone_list,
+        speaker_id=np.array([speaker_id], dtype=np.int64).reshape(-1),
+        stereo=stereo,
+        sample_rate=sample_rate,
+        volume=volumeScale,
+    )
+
+
+def accent_phrase(
+    engine: SynthesisEngineBase,
+    audio_file: Optional[IO],
+    kana: str,
+    speaker_id: int,
+    normalize: int,
+) -> List[AccentPhrase]:
+    _lazy_init()
+    f0, phonemes = extract_feature(audio_file, kana)
+    timed_phonemes = frame_to_second(deepcopy(phonemes))
+    accent_phrases, _ = parse_kana(kana, False)
+
+    phrase_info = []
+    for ((s, e, p), (ts, te, tp)) in zip(phonemes, timed_phonemes):
+        if p not in unvoiced_mora_phoneme_list:
+            clip = f0[_resample_ts(s) : _resample_ts(e)]
+            clip = clip[clip != 0]
+            pitch = np.average(clip) if len(clip) != 0 else 0
+        else:
+            pitch = 0
+        pitch = 0 if np.isnan(pitch) else pitch
+        length = float(te) - float(ts)
+        phrase_info.append(PhraseInfo(pitch, length, p))
+
+    if normalize:
+        f0_avg = _no_nan(np.average(f0[f0 != 0]))
+        predicted_avg = _predict_avg_pitch(engine, kana, speaker_id)
+        normalize_scale = predicted_avg / f0_avg
+        for p in phrase_info:
+            p.pitch = p.pitch * normalize_scale
+
+    idx = 1
+    for phrase in accent_phrases:
+        for mora in phrase.moras:
+            if mora.consonant is not None:
+                mora.pitch = (phrase_info[idx].pitch + phrase_info[idx + 1].pitch) / 2
+                mora.consonant_length = phrase_info[idx].length
+                mora.vowel_length = phrase_info[idx + 1].length
+                idx += 2
+            else:
+                mora.pitch = phrase_info[idx].pitch
+                mora.vowel_length = phrase_info[idx].length
+                idx += 1
+        if phrase_info[idx].phoneme == "sp":
+            phrase.pause_mora = Mora(
+                text="、",
+                consonant=None,
+                consonant_length=None,
+                vowel="pau",
+                vowel_length=phrase_info[idx].length,
+                pitch=0,
+            )
+            idx += 1
+
+    return accent_phrases
+
+
+def extract_feature(audio_file: Optional[IO], kana: str):
+    sr, wave = wavfile.read(audio_file)
+    if len(wave.shape) == 2:
+        wave = wave.sum(axis=1) / 2
+
+    f0 = extract_f0(wave, sr, 256 / 24000 * 1000)
+
+    julius_wave = resample(wave.astype(np.int16), JULIUS_SAMPLE_RATE * len(wave) // sr)
```

Comment:
[QUESTION] Is audio query always 16-bit integer PCM? If not, you have to renormalize the wave.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.io.wavfile.read.html

Reply:
I haven't yet encountered an audio file in other formats but yes, I agree that we should add a boundary check just in case.